### PR TITLE
Add Exemplars specification.

### PR DESCRIPTION
**Pull request by songy23**

### CODE REVIEW

File: stats/Export.md [hunk 1]
```diff
@@ -15,7 +15,8 @@ The library SHOULD provide support for multiple types of Aggregations:
 type.
 * `LastValueDataDouble` and `LastValueDataInt64`: data generated for a `LastValue` aggregation based 
 on the `Measure` type.
-* `DistributionData`: data generated for a `Distribution` aggregation.
+* `DistributionData`: data generated for a `Distribution` aggregation. If there is a histogram in 
+the `Distribution`, `DistributionData` MAY contain a list of `Exemplar`s.
```
Comment (ramonza) [inline stats/Export.md]: I thought we were moving away from the "specalese" capitalization of random words =)

Comment (songy23) [reply stats/Export.md]: I see, updated.

File: stats/Record.md [hunk 2]
```diff
@@ -67,6 +69,7 @@ private static final MeasureLong RPC_BYTES_SENT =
 MeasurementMap measurementMap = new MeasurementMap();
 measurementMap.put(RPC_LATENCY, 10.3);
 measurementMap.put(RPC_BYTES_SENT, 124);
+measurementMap.withSpanContext(spanContext);
```
Comment (sebright) [inline stats/Record.md]: Should the Java implementation support using the current span?  In that case, I think the no-arg `record` method should  add the spanContext implicitly.

Comment (songy23) [reply stats/Record.md]: Do you mean adding another no-arg `recordWithCurrentSpanContext()` method? I think for the existing no-arg `record()` won't interact with spanContext, to be consistent with previous behavior?

Comment (ramonza) [reply stats/Record.md]: What is the harm in sending exemplars by default? Why do we need to remain consistent with the previous behavior?

If we do, I would recommend instead that we configure exemplar support on the exporter (those that support it, which is just Stackdriver for now). 

I feel quite strongly that we shouldn't require people to change their instrumentation to get exemplars if we can avoid it.

Comment (songy23) [reply stats/Record.md]: I don't object to make it default to record again current span context, though I'm wondering will there be cases that user would only want histograms, but not linked to trace spans? If so, I think we should at least provide a method that records stats against no span context.

**Comment by songy23** [comment]
PTAL. Per our meeting today, we would like to have a more generic record() method that takes a `Map<String, String>` as the attachment of `Exemplar` to avoid having dependency on Tracing APIs. The client code should take care of encoding their exemplar attachments (like TraceId, SpanId) to plain string.

File: stats/Record.md [hunk 3]
```diff
@@ -69,7 +69,15 @@ private static final MeasureLong RPC_BYTES_SENT =
...
 measurementMap.record();  // reads context from thread-local.
+
+// Another example on recording against sampled SpanContext.
+SpanContext spanContext = tracer.getCurrentSpan().getContext();
+if (spanContext.getTraceOptions().isSampled()) {
+  Map<String, String> attachments = new HashMap<>();
+  // Client code needs to take care of encoding.
+  attachments.put("TraceId", encode(spanContext.getTraceId));
+  attachments.put("SpanId", encode(spanContext.getSpanId));
+  measurementMap.record(attachments);
```
Comment (sebright) [inline stats/Record.md]: I'm not sure if this makes sense, but another way to handle explicit propagation is to pass the whole gRPC context to `record`.  Then the two `record` methods would behave more similarly.  They would both allow the OpenCensus implementation to control how the exemplars are added, rather than pushing that logic into the client code.  I'm not sure how that would work in all languages, though, since opencensus-specs doesn't require a context, like the gRPC context or Go context, that references the SpanContext and TagContext and can be passed around.

File: stats/Record.md [hunk 4]
```diff
@@ -68,5 +70,14 @@ MeasurementMap measurementMap = new MeasurementMap();
...
 measurementMap.put(RPC_BYTES_SENT, 124);
 measurementMap.record();  // reads context from thread-local.
+
+// Another example on recording against sampled SpanContext.
+SpanContext spanContext = tracer.getCurrentSpan().getContext();
+if (spanContext.getTraceOptions().isSampled()) {
+  Map<String, String> attachments = new HashMap<>();
+  // Client code needs to take care of encoding.
+  attachments.put("TraceId", encode(spanContext.getTraceId));
+  attachments.put("SpanId", encode(spanContext.getSpanId));
```
[the remark this answers is not in the document]
Comment (songy23) [reply stats/Record.md]: I think that's an implementation detail of the `encode()` method. Client code should decide whether or not they want the lowercase base16 encoding of TraceId/SpanId (there may also be clients that need some other encoding format).

File: stats/Record.md [hunk 5]
```diff
@@ -68,5 +70,14 @@ MeasurementMap measurementMap = new MeasurementMap();
...
+  measurementMap.record(attachments);
```
[the remark this answers is not in the document]
Comment (songy23) [reply stats/Record.md]: Yes, done.

File: stats/Record.md (hunk 3 above)
Comment (songy23) [reply stats/Record.md]: There could be other exemplar attachments that are not from the current context, like [plain string literals](https://cloud.google.com/monitoring/api/ref_v3/rest/v3/TimeSeries#exemplar). I think our goal here is to make the record() method as generic as possible, and restricting it to take only the current context seems to be against that goal.

File: stats/Record.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (ramonza) [reply stats/Record.md]: There needs to be agreement between the gRPC code and the Stackdriver exporter. Maybe we should make it a `Map<String, Object>` instead and pass the whole `spanContext`?

[the remark this answers is not in the document]
Comment (songy23) [reply stats/Record.md]: This makes sense to me, by doing that client code doesn't need to worry about encoding too. Should we change the type of [exemplar.attachment](https://github.com/census-instrumentation/opencensus-proto/pull/62/files#diff-cbf22e3bb90f48f62fe55b894a9ee40cR236) to map<string, Any> to match this?

[the remark this answers is not in the document]
Comment (ramonza) [reply stats/Record.md]: Sorry to go in circles on this but we added the Map because we didn't want an explicit dependency from the stats API to the trace API. Maybe we should have instead just made it `Object`/`interface{}` and let the stats API just pass it through? To put it another way, what purpose is the key in the map serving?

That way only gRPC and the exporters need to know about the actual type. The stats API can just treat it as an opaque pointer/object and pass it along. Then `exemplar.attachment` in the proto can go back to being just `Any`.

[the remark this answers is not in the document]
Comment (songy23) [reply stats/Record.md]: > To put it another way, what purpose is the key in the map serving?

FWIW we want a map here because there may be different types of Exemplar attachments for a single histogram bucket. For example, we can have both a set of dropped labels and a SpanContext associated with a histogram bucket. In that case, map keys are needed to differentiate attachments of different types (like a JSON object).

> That way only gRPC and the exporters need to know about the actual type. The stats API can just treat it as an opaque pointer/object and pass it along.

I think we can achieve the same goal with Map<String, Object>. The stats API doesn't need to know the actual type of the map value and can just pass it along.

File: stats/Record.md [hunk 6]
```diff
@@ -44,6 +44,8 @@ measure name instead of the `Measure`.
 
 Implementations MAY define a `MeasurementMap` which describes a set of data points to be collected
 for a set of Measures. Adding this functionality may improve the efficiency of the record usage API.
+Besides, when recording Measurements, `MeasurementMap` should optionally take a map of string-object
+pairs as the attachment of exemplar.
```
Comment (dinooliva) [inline stats/Record.md]: This is more generic than originally discussed (i.e. Map<String, String>) but I think makes sense. Let's discuss it in the cross-language sync in case non-gc'd languages may have concerns about object lifetimes.

File: stats/Record.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (sebright) [reply stats/Record.md]: I think Object is too generic, because exporters won't be able to pass it along without knowing what it represents, since it's not possible to serialize unknown Objects.  Using String values gives exporters the flexibility to either look for specific keys that they understand, or export the whole map.

File: stats/Record.md (hunk 3 above)
Comment (sebright) [reply stats/Record.md]: Then it seems like we should try to make the implicit propagation version of `record` more generic, too, though I'm not sure how that would work.  Maybe both `record` methods should take a `Map<String, String>`, and the implicit propagation version could merge the input map with the keys-value pairs representing the current span.

File: stats/Record.md (hunk 6 above)
Comment (songy23) [reply stats/Record.md]: Per our discussion today I've reverted the change to use Map<String, String> instead.

File: stats/Record.md [hunk 7]
```diff
@@ -44,11 +44,15 @@ measure name instead of the `Measure`.
 
 Implementations MAY define a `MeasurementMap` which describes a set of data points to be collected
 for a set of Measures. Adding this functionality may improve the efficiency of the record usage API.
+Besides, when recording Measurements, `MeasurementMap` should optionally take a map of string
```
[the remark this answers is not in the document]
Comment (songy23) [reply stats/Record.md]: Done.

File: stats/Record.md [hunk 8]
```diff
@@ -44,11 +44,15 @@ measure name instead of the `Measure`.
 
 Implementations MAY define a `MeasurementMap` which describes a set of data points to be collected
 for a set of Measures. Adding this functionality may improve the efficiency of the record usage API.
+Besides, when recording Measurements, `MeasurementMap` should optionally take a map of string 
+key-value pairs as the attachment of exemplar.
```
[the remark this answers is not in the document]
Comment (songy23) [reply stats/Record.md]: I think we defined `Exemplar` as `value` + `timestamp` + `attachments` in https://github.com/census-instrumentation/opencensus-proto/pull/62. `value` is included in `measurement`, `timestamp` is added by the library, so we need to pass in the `attachments`, which is the string map.

[the remark this answers is not in the document]
Comment (songy23) [reply stats/Record.md]: I see, would it be better to rephrase this to "MeasurementMap should optionally take a map of string key-value pairs to record exemplar"?

[the remark this answers is not in the document]
Comment (dinooliva) [reply stats/Record.md]: That works for me - nit: "to record an exemplar" or "to record exemplars"

[the remark this answers is not in the document]
Comment (songy23) [reply stats/Record.md]: Yep, updated.

File: stats/Record.md [hunk 9]
```diff
@@ -44,11 +44,15 @@ measure name instead of the `Measure`.
...
 for a set of Measures. Adding this functionality may improve the efficiency of the record usage API.
+Additionally, when recording Measurements, `MeasurementMap` should optionally take a map of string 
+key-value pairs to record an exemplar.
 
 ## Recording Stats
 
 Users should record Measurements against a context, either an explicit context or the implicit 
 current context. Tags from the context are recorded with the Measurements if they are any.
+Note that there is no implicit recording for exemplars. If you want to record a `Measurement`
+against an exemplar, you have to explicitly pass the exemplar attachments (a string-string map).
```
Comment (dinooliva) [inline stats/Record.md]: You should word this similarly (i.e. getting rid of the term attachments).

File: stats/Record.md [hunk 10]
```diff
@@ -68,5 +72,14 @@ MeasurementMap measurementMap = new MeasurementMap();
...
 measurementMap.record();  // reads context from thread-local.
+
+// Another example on recording against sampled SpanContext.
+SpanContext spanContext = tracer.getCurrentSpan().getContext();
+if (spanContext.getTraceOptions().isSampled()) {
+  Map<String, String> map = new HashMap<>();
+  // Client code needs to take care of encoding.
+  map.put("TraceId", encode(spanContext.getTraceId()));
+  map.put("SpanId", encode(spanContext.getSpanId()));
+  measurementMap.record(tagContext, map);
```
Comment (adriancole) [inline stats/Record.md]: can someone mention how an existing metrics system might handle this data as it is inherently unlimited cardinality. For example, is there an exemplar feature somewhere? I get where this is going, but bottom-up I'm kindof guessing.

Comment (bogdandrutu) [reply stats/Record.md]: @adriancole here are some examples of data we want to record:
1) TraceId/SpanId this will allow the backend to display on their UI information about traces within a percentile (usually) (e.g. "Here are 99%tile traces").
2) Dropped tags. Let's assume that in the tag map/set/context we have 20 unique tags and we break down stats only by 4 of them. We record the rest of 16 as dropped tags. This can be used as example of requests tags for different percentiles.

The first example is what we call correlation between metrics and traces.
